yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

```diff
@@ -256,6 +256,12 @@ private[spark] class ApplicationMaster(
 
   private def runDriver(securityMgr: SecurityManager): Unit = {
     addAmIpFilter()
+
+    // This must be done before SparkContext is initialized, since the CoarseGrainedSchedulerBackend
+    // is started at that time. That is what schedules the re-logins. It is scheduled only if the
+    // principal is actually setup. So we make sure it is available.
+    SparkHadoopUtil.get.setPrincipalAndKeytabForLogin(
+      System.getenv("SPARK_PRINCIPAL"), System.getenv("SPARK_KEYTAB"))
```

Comment:
yeah we should use configs whenever possible.  env variables are mostly old and haven't been cleaned up yet.

Configs also give you the option of setting these things in your spark conf file rather then having to pass them in every time.   Make sure to name them something obvious and yarn specific.

History server keytab configs: http://spark.apache.org/docs/1.1.0/monitoring.html